simtools/data_model/metadata_collector.py

```diff
@@ -0,0 +1,292 @@
+            )
+        try:
+            for document in _input_meta["product"]["document"]:
+                self._fill_context_sim_list(top_level_dict["context"]["sim"]["document"], document)
+        except KeyError:
+            top_level_dict["context"]["sim"].pop("document")
+
+    def _fill_product_meta(self, product_dict):
+        """
+        Fill metadata for data products fields. If a schema file is given for the data products,
+        try and read product:data:model metadata from there.
+
+        Parameters
+        ----------
+        product_dict: dict
+            Dictionary describing data product.
+
+        Raises
+        ------
+        KeyError
+            if relevant fields are not defined in top level metadata dictionary.
+
+        """
+
+        product_dict["id"] = self.args_dict.get("activity_id", "UNDEFINED_ACTIVITY_ID")
+        self._logger.debug(f"Reading activitiy UUID {product_dict['id']}")
+
+        product_dict["data"]["category"] = "SIM"
+        product_dict["data"]["level"] = "R0"
+        product_dict["data"]["type"] = "service"
+        _schema_dict = (
+            gen.collect_data_from_yaml_or_dict(
+                in_yaml=self.args_dict.get("schema", None), in_dict=None, allow_empty=True
+            )
+            or {}
+        )
+        product_dict["data"]["model"]["name"] = _schema_dict.get("name", "simpipe-schema")
+        product_dict["data"]["model"]["version"] = _schema_dict.get("version", "0.0.0")
+        product_dict["format"] = self.args_dict.get("output_file_format", None)
+        product_dict["filename"] = str(self.args_dict.get("output_file", None))
+
+    @staticmethod
+    def _fill_association_id(association_dict):
+        """
+        Fill association id from site and telescope class, type, subtype.
+
+        Parameters
+        ----------
+        association_dict: dict
+            Association dictionary.
+
+        """
+        for association in association_dict:
+            try:
+                association["id"] = names.simtools_instrument_name(
+                    association["site"],
+                    association["class"],
+                    association["type"],
+                    association.get("subtype", None),
+                )
+            except ValueError:
+                association["id"] = None
+
+    def _fill_activity_meta(self, activity_dict):
+        """
+        Fill activity (software) related metadata
+
+        Parameters
+        ----------
+        activity_dict: dict
+            Dictionary for top-level activitiy metadata.
+
+        Raises
+        ------
+        KeyError
+            if relevant fields are not defined in top level metadata
+            dictionary
+
+        """
+
+        activity_dict["name"] = self.args_dict.get("label", None)
+        activity_dict["start"] = datetime.datetime.now().isoformat(timespec="seconds")
+        activity_dict["end"] = activity_dict["start"]
```

Comment:
Can you help me with this comment? I don't understand it. Are you talking about `dict_high`?

Reply:
Sorry, no I get your comment. Is it like that?
```
>>> activity_dict = {}
>>> activity_dict['start'] = "10.00"
>>> activity_dict['end'] = activity_dict['start']
>>> print(activity_dict['start'], activity_dict['end'])
10.00 10.00
>>> activity_dict['end'] = "12.00"
>>> print(activity_dict['start'], activity_dict['end'])
10.00 12.00
```

Reply:
I was actually more commenting on the fact that `activity_dict["end"] = activity_dict["start"]` and I didn't see where you set the `end` to be a different value later (assuming the processing took time).

My comment on them not being a copy is simply wrong. I am sure at some point I ran into this issue of a variable not being a copy, but this is clearly not the case so ignore it.

Reply:
The activity is the running of the simtools tool. I assume that we write efficient code and which executes instantaneously :-) 

Do you think it is worth adding code to be precise here? Set the 'end' at the actually end of this activity? Some tools take longer.

Reply:
Some tools definitely take longer, especially those which derive parameters. I would therefore suggest to add code that does it (can be a separate PR of course). Otherwise we are just complying with the data model but not providing meaningful data in it..